### PR TITLE
Feature/6 0/refinery in http service

### PR DESCRIPTION
**Pull request by chfsx**
Dear Technical Board

This PR proposes using the refinery in the HTTP service. It includes the following changes, some of which are still under discussion:

- Separation of the global HTTp service : A distinction is made between raw() and wrapper(). In the future, only the wrapper should be used, but currently many services still access the entire `RequestInterface`. 
- The wrappers offer easy ways to access $_GET, $_POST and $_COOKIE using `transformations` of the refinery (see https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-eda6070f33899930ba976bfa6b85334cR21 )

This is the part we should discuss:
- The PR also proposes drop-in replacements for the superglobal $_GET, $_POST, $_REQUEST and $_COOKIE. These classes enclose the superglobals and use the refinery internally when accessing content (see https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-2f6f2df4c9612331e495a008236e315bR12 ). This implementation is certainly worth discussing, which also needs some adjustments in ILIAS because they expect native arrays (e.g. https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-411e64d40db5eeb3d6a82486fdc5e607R526 ).
- In some places the superglobals in ILIAS were simply overwritten, this must not be from my point of view, I commented out these places (otherwise also the DropIn Replacements do not work) (e.g. https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-c96313dda4795745e3af54ee394b156eL181 ).

Please also contact me with questions, I will be happy to give you a walk-through if you wish.

### CODE REVIEW

**Comment by alex40724** [comment]
@chfsx Thanks a lot for the PR. One question: I am missing instructions / README changes that instruct developers how to finally use this. Is it intended to add this?

**Comment by chfsx** [comment]
@alex40724 yes of course, but that depends a bit from what parts of this PR will finally be available in trunk or even release_6. the wrappers I think are uncritical. but what about the drop-in-replacements? and what will be the roadmap to get rid of them? this is a part the TB should decide

File: src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        if (!$this->offsetExists($offset)) {
+            return null;
+        }
+
+        if (is_array($this->raw_values[$offset])) {
+            $this->factory->to()->dictOf($this->factory->to()->string())->transform((array) $this->raw_values[$offset]);
+        }
+
+        return $this->factory->to()->string()->transform((string) $this->raw_values[$offset]);
```
Comment (mjansenDatabay) [inline src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php:60]: Maybe I did not get it (yet), but do we need this call if `$this->raw_values[$offset]` is an array?

File: src/HTTP/Wrapper/ArrayBasedRequestWrapper.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        $this->raw_values = $raw_values;
+    }
+
+
+    /**
+     * @inheritDoc
+     */
+    public function retrieve(string $key, Transformation $transformation)
+    {
+        if (!$this->has($key)) {
```
Comment (mjansenDatabay) [inline src/HTTP/Wrapper/ArrayBasedRequestWrapper.php:37]: I suggest to use a `OutOfBoundsException ` instead.

File: src/HTTP/Wrapper/WrapperFactory.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    public function post() : RequestWrapper
+    {
+        return new ArrayBasedRequestWrapper($this->request->getParsedBody());
+    }
+
+
+    public function cookie() : RequestWrapper
+    {
+        return new ArrayBasedRequestWrapper($_COOKIE);
```
Comment (mjansenDatabay) [inline src/HTTP/Wrapper/WrapperFactory.php]: Is there any reason why you did not use `$this->request->getCookieParams()` here?

**Review by mjansenDatabay — COMMENTED** [review]
Thanks for this PR @chfsx !

At a first glance, it was a bit hard (for me) to understand all the `lazy loading reflection and caching stuff`, but I think I finally got it.

I personally like that you removed the registration of the different HTTP components from the DIC and only `http` as a global service is left, at least for the most common requests where `ilInititialisation::initHTTPServices` is called.

Furthermore I like the drop-in-replacement approach via `DirectArrayAccessDropInReplacement` which could be used to make the super globals immutable in future. Using a hard coded `string() Refinery Transformation` there is not very flexible, but this is fine for me as long as we do not comprehensively use the `HTTP Service` in our code base.

I added some minor remarks as inline comments.

I'll most probably provide more feedback soon, but that's it for the first review ...

File: src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php (hunk 1 above)
Comment (chfsx) [reply src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php:60]: there should be a return in line 57:

`return $this->factory->to()->dictOf($this->factory->to()->string())->transform((array) $this->raw_values[$offset]);`

File: src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+
+    /**
+     * @inheritDoc
+     */
+    public function offsetSet($offset, $value) : void
+    {
+        // throw new \LogicException("Modifying global Request-Array such as \$_GET is not allowed!");
+        $this->raw_values[$offset] = $value;
```
Comment (chfsx) [inline src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php:70]: I'll use the `OutOfBoundsException` here as well, thanks!

File: src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php (hunk 1 above)
Comment (mjansenDatabay) [reply src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php:60]: Exactly ;-)

**Review by klees — CHANGES_REQUESTED** [review]
Dear @chfsx,

we are sorry that it tooks us so long to give feedback here, like most ILIAS developers we were incredibly busy with 6.0. We hope you still remember what was going on and can finish the work on this PR. We deem the introduction of the Refinery to the HTTP service to be a huge step for ILIAS towards a tightened input security. Thanks for working on this, please get back to us if you need additional funding to finish.

The general pattern for the access to request data via the Refinery is implemented as dicussed in the input security paper. We would like to boost the usage of that new pattern and phase out direct access to the request super globals and thus have some change requests to make this happen. We are not sure about the introduction of `DirectArrayAccessDropInReplacement` for the super globals and would like to discuss pros and cons of that move.

We also have some questions and requests that are more about the internal workings of your service, which is not strictly what this PR or our initial request was about. We still are interested in these because HTTP is a very central piece of infrastructure of a web application =)

Please answer the following questions:
- [x] ilPropertyFormGUI: Can we really remove [this](https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-411e64d40db5eeb3d6a82486fdc5e607L503) chunk of code? Why don't we need it anymore?
- [x] ReflectionClass: What's the purpose of [`HTTP\Services::getWithMultipleArguments`](https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-e50f175337eeff55968c5b4b2650f8c8R153)? Using the `ReflectionClass` is (deemed to be) a performance impact. Why do we take it here? We know the classes that we want to instantiate.
- [x] DirectArrayAccessDropInReplacement: What is the purpose or added benefit of `DirectArrayAccessDropInReplacement`? Using the `Refinery` internally won't do anything to make this more secure, at least regarding the current implementation. Transforming the contents to string basically is a non-operation from the perspective of the result, the contents in the request super globals are strings anyway.

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] DI: Please do not remove the dependency injection of `GlobalHttpState`. This will make testing and reuse of the class harder. What's the benefit of creating dependencies inline?

Please implement the following changes:
- [ ] Wrapper/Raw: In current proposal you suggest to add a `wrapper` method to the `GlobalHttpState`, which will give access to the methods that use the `Refinery` for data retrieval. These should become the new normal for almost all of the ILIAS code. The other methods, like `request`, on the `GlobalHttpState` on the other hand provide access to the raw request objects, which mostly should not be to used anymore. Thus we suggest to change the `GlobalHttpState` so that `request` and `cookieJar` return the objects that use the `Refinery` for access and only make the raw objects accessible behind `raw`. Maybe we could even give a name that hints at the unsafeness even more, like  `getUnsafeRawObjects`. An update of the existing usage should be straight forward.
- [ ] HTTPService: With the changes here, `HTTPServices` seems to have become an odd name. Please rename to `RawHTTPObjects`.
- [ ] RequestWrapper: This should be pretty generic for all kinds of inputs. Rename it to something like `KeyValueInput` and move it to the top-level of the Refinery. The same goes for `ArrayBasedRequestWrapper`, which could be renamed to `ArrayInput` and also be moved to the `Refinery`.
- [ ] Dicto: Add some Dicto rules that flag usages of the request super globals and `getUnsafeRawObject`.
- [ ] Tests: Please add automated tests that at least cover the core features of the HTTP library.

If you have any questions, feel free to ask here or contact us to make an appointment via Skype. Thanks again for taking care about this central pillar of our software!

Best regards!
Richard on the behalf of the Technical Board of the ILIAS society

**Comment by kergomard** [comment]
@chfsx : In our tb meeting we stumbled over this open PR belonging to a offer from last year. We would kindly ask you to react to the above feedback from the TB. Thank you very much!

**Comment by chfsx** [comment]
Thanks for the review and no problem that it takes a little longer (it took me a long time now, too :-) )

with pleasure I will go into your points:

>  ilPropertyFormGUI: Can we really remove this chunk of code? Why don't we need it anymore?

this was to check if post_max_size exceeded, but this leads to an error anyway.

> ReflectionClass: What's the purpose of HTTP\Services::getWithMultipleArguments? Using the ReflectionClass is (deemed to be) a performance impact. Why do we take it here? We know the classes that we want to instantiate.

This was to handle the internal instances a bit easier, but I can remove this as well

> DirectArrayAccessDropInReplacement: What is the purpose or added benefit of DirectArrayAccessDropInReplacement? Using the Refinery internally won't do anything to make this more secure, at least regarding the current implementation. Transforming the contents to string basically is a non-operation from the perspective of the result, the contents in the request super globals are strings anyway.

Its correct that this doesn't bring a direct improvement of security. But it's a way to handle code which accesses $_GET directly in the future. With this implementation all accesses to $_GET go through this implementation, it will be easy to prevent from doing this in the future (which should be done IMO). I highly recommend to use this.

> DI: Please do not remove the dependency injection of GlobalHttpState. This will make testing and reuse of the class harder. What's the benefit of creating dependencies inline?

I'll adapt this in the next iteration, no special intention behind this change

I will be happy to adapt the desired customizations in the next iteration, thanks for your feedback!

**Comment by chfsx** [comment]
Due to a lot of stuff which changed since the original PR (to release_6) I'll open a new PR for trunk asap. In the new PR I'd like to discuss some points with you, thanks in advance

**Pull request closed without merging**